File: 2981-find-longest-special-substring-that-occurs-thrice-i/2981-find-longest-special-substring-that-occurs-thrice-i.py
class Solution:
    def ok(self, s: str, md: int) -> bool:
        vis = [0] * 26  
        l = 0

        for r in range(len(s)):
            if s[r] != s[l]:
                l = r 
            if r - l + 1 >= md:
                vis[ord(s[r]) - ord('a')] += 1
            if vis[ord(s[r]) - ord('a')] > 2:
                return True
        
        return False

    def maximumLength(self, s: str) -> int:
        st, ed = 1, len(s)
        cur = -1

        while st <= ed:
            md = (st + ed) // 2
            if self.ok(s, md):
                cur = md
                st = md + 1
            else:
                ed = md - 1

        return cur

File: 2981-find-longest-special-substring-that-occurs-thrice-i/2981-find-longest-special-substring-that-occurs-thrice-i.cpp
class Solution {
public:
    bool ok(string &s,int md){
        vector<int>vis(26);
        int l = 0;
        for(int r = 0 ; r < s.size();r++){
            if(s[r] != s[l]) l = r;
            if(r-l+1>=md)vis[s[r]-'a']++;
            if(vis[s[r]-'a']>2)return true;
        }
        return false;
    }
    int maximumLength(string s) {
        int st = 1,ed = s.size() ,cur = -1,md;
        while(st<=ed){
            md = (st+ed)/2;
            if(ok(s,md)){
                cur = md;
                st = md+1;
            }
            else ed = md-1;
        }
        return cur;
    }
};